Add support for Plymouth boot splash

FlyingWaffleDev opened the pull request:
Since udev is now included by default when using genkernel, adding support for Plymouth wasn't too difficult.  As far as I can tell, this was also the last major feature that genkernel-next supported that genkernel does not.

I tested it out on my system (following Sakaki's EFI Guide) and it appears to work just fine!

I followed gk-next as a guidline when writing this, but everything has been coded from scratch.  If there is anything that needs changing please let me know.

heroxbd commented:
Please at least replace the spaces with tabs, keeping the same style.

FlyingWaffleDev commented:
> Please at least replace the spaces with tabs, keeping the same style.

I have my vimrc set to automatically substitute tabs with spaces, so I didn't notice.
Hopefully I haven't missed any.

heroxbd commented:
> > Please at least replace the spaces with tabs, keeping the same style.
> 
> I have my vimrc set to automatically substitute tabs with spaces, so I didn't notice.
> Hopefully I haven't missed any.

Thanks!  The actual code review is beyond my expertise.  Good luck and have fun with Gentoo!

FlyingWaffleDev commented:
I had remnants of an old initramfs that were screwing with my testing of the additions, but I've cleared that up now.  So at this point the plymouth support 'works' except for the theme.  I'm trying to hunt down the problem, but as far as I can tell there is no significant difference from how genkernel-next does things; so I have no idea why plymouth themes aren't being called correctly.  I'll keep working on it; if anyone has any ideas please let me know.

FlyingWaffleDev commented:
Everything works properly now!  Only remaining quibble is a "sh: 1: unknown operand" that pops up before the splash theme and also shows up once or twice in log files.

codedmart commented:
For what it is worth I have tested this and all is working quite well.

niyumard commented:
First of all thank you very much @FlyingWaffleDev, this is exactly the feature I've been looking forward to, and then, how about deleting the lines pertaining to fbsplash/splashutils in genkernel.conf since it's not in the package tree any more?

Whissi commented:
Just an ACK that we have noticed this PR. Plymouth support is def. on my list for genkernel-4. But it will take some time to review and have a look.

maffblaster commented:
Tracked in Gentoo here: https://bugs.gentoo.org/753617

FlyingWaffleDev commented:
Just tested after updating to latest git version from version 4.1.2 and resolving merge conflicts; everything still works as expected.

hijackeel commented:
I've also tested this, and it works.

I made my own fork of buildkernel that uses it: https://github.com/hijackeel/buildkernel-hjkl
Overlay fork with ebuild: https://github.com/hijackeel/sakaki-tools-hjkl

Afterwards I saw that @FlyingWaffleDev had already made his own fork of buildkernel with ebuild. :/

maffblaster commented:
Hmm... Seems like it would be nice to have movement on this item. I'll see what's possible. This discussion brought me here: https://wiki.gentoo.org/wiki/Talk:Plymouth#genkernel-next_removal

hijackeel commented:
I've tested this with genkernel-4.2.6-r2 and it works.

robbat2 commented:
@FlyingWaffleDev hi, conceptually it looks good, but I'd like for you to clean up the series of commits, esp adding signed-off-by on the new final commits.

Will leave other comments at code lines

robbat2 commented on defaults/initrd.scripts:
```diff
@@ -1966,15 +2043,27 @@ openLUKS() {
-				crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+				if [ "${PLYMOUTH}" = '1' ]
+				then
+					ply_cmd="${ply_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
```
can you document what this is doing? I'm surprised to see plymouth changes touching LUKS

FlyingWaffleDev replied:
Plymouth doesn't actually touch anything; but the actual cryptsetup command to unlock the LUKS partition is called by plymouth.  To do this, the command is built out (ply_cmd; the gpg call requires specifying stdin and not the standard password box) and written to a file.  This file is then read by plymouth to run the command (called in crypt_filter), pass the typed password from plymouth via stdin and unlock the LUKS partition.

FlyingWaffleDev replied:
Should I provide commented documentation in the code?

FlyingWaffleDev replied:
I've included documentation on the specific gpg options used in my new commit (squashed).

robbat2 commented on gen_initramfs.sh:
```diff
@@ -1284,6 +1285,137 @@ append_splash() {
+		fi
+
+		print_info 1 "$(get_indent 2)>> Copying ${mod}..."
+		cp -ax --parents "${mod}" "${TDIR}" || gen_die "Failed to copy '${mod}'!"
```
Why not use the normal modules functionality? This is instead causing the module to be embedded into the initramfs twice potentially (once in the drm fragment, and again in the module fragment)

FlyingWaffleDev replied:
Back when I originally wrote this I didn't understand the append_modules function properly (and honestly I still don't).  So if I understand you correctly, I don't actually need any of the module/firmware appending code in append_drm() at all since append_modules() will pull it all in automatically?

FlyingWaffleDev replied:
I feel really silly now; I did some testing and the existing --firmware and --all-ramdisk-modules options are sufficient for plymouth to function.  My append_drm function apparently didn't anything at all either, my kernel size didn't change at all after removing it.

So, that resolves these change requests and also makes building initramfs with plymouth significantly faster.

robbat2 commented on gen_initramfs.sh:
```diff
@@ -1284,6 +1285,137 @@ append_splash() {
+
+		# check that firmware files exist, then copy them (modinfo may list deprecated firmware)
+		fws=( $(get_firmware_files "${mod}") )
+		for fw in "${fws[@]}"
```
similar problem here, now you're embeddeding firmware twice.

FlyingWaffleDev replied:
Haven't looked at this code in a while, but I believe the idea was to make sure the required modules/firmware for plymouth were included even when the --firmware and --all-ramdisk-modules options were not selected.  Would it be a better idea to make plymouth=true also set firmware and all-ramdisk modules in gen_determineargs.sh?  (Or at least report a gen_die error message if they aren't set?)

robbat2 commented on gen_initramfs.sh:
```diff
@@ -1284,6 +1285,137 @@ append_splash() {
+
+get_firmware_files() {
+	local kmod="${1}"
+	modinfo --set-version="${KV}" -F firmware "${kmod}" || gen_die "Failed to get modinfo for ${kmod}!"
```
TIL this functionality!

robbat2 commented on gen_initramfs.sh:
```diff
@@ -1998,11 +2130,17 @@ create_initramfs() {
 	append_data 'xfsprogs' "${XFSPROGS}"
 	append_data 'zfs' "${ZFS}"
 
+	if isTrue "${PLYMOUTH}"
```
why not just `append_data 'drm' "${PLYMOUTH}"`?

FlyingWaffleDev replied:
Ah, I got a bit confused by the surrounding "if isTrue" functions.

FlyingWaffleDev replied:
Actually, looking at it again, I'm doing the same thing as ZFS.  Both plymouth and ZFS have an initial append_data() call, then do an "if isTrue" for additional append_data() calls.  Should I not be doing this?

FlyingWaffleDev committed:
Plymouth support for genkernel (squashed).

Signed-off-by: FlyingWaffle <flyingwaffle@pm.me>

FlyingWaffleDev commented:
Tried rebasing to squash the commits, holy crap that was frustrating.  Ended up just doing a hard reset to master and applying everything as a single patch then force pushed it.

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -2068,8 +2146,19 @@ openLUKS() {
+                printf '#!/bin/sh\n%s\n' "${ply_cmd}" > "${ply_cmd_file}"
+                run chmod 500 "${ply_cmd_file}"
+            fi
+            crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}" "${ply_cmd_file}"
```
Please create separate `crypt_filter_ply()` (or smth. similar) and place it inside condition block above. Or declare `local ply_cmd_file=""` near of `local ply_cmd=""`. Also use tabs instead of spaces.

reagentoo replied:
What happens if the `LUKS_KEY` is defined? Will `plymouth ask-for-password` show a password entry dialog and skip right away?

FlyingWaffleDev replied:
Will push the changes once I get home.

I'm not quite sure what you're asking about the LUKS_KEY.  If it is a gpg key, then plymouth shows a password entry dialog and you have to enter your key decryption password correctly before it continues.  I have not tested the case where a non-gpg, non-password-protected key is used, but I presume you could just hit enter on the empty password field to continue.

reagentoo replied:
I think it is necessary to prevent handling by plymouth if non-gpg key is used. Show a password dialog only for missing key and gpg-key.

reagentoo replied:
Please ignore my first comment ab. `crypt_filter_ply()`. I suggest the following `crypt_filter()`:
```
crypt_filter() {
	[ -z "${LUKS_KEY}" -o "$(echo ${LUKS_KEY} | grep -o '.gpg$')" = ".gpg" ]
	local ply_ask=$?

	if [ "${CRYPT_SILENT}" = '1' ]
	then
		eval run ${1}
	elif [ "${PLYMOUTH}" = '1' ] && plymouth_running && [ ply_ask -eq 0 ]
	then
		local ply_cmd_file="$(mktemp -t 'ply_cmd.XXXXXX' 2>/dev/null)"
		printf '#!/bin/sh\n%s\n' "${1}" > "${ply_cmd_file}"
		run chmod 500 "${ply_cmd_file}"
		run plymouth ask-for-password --prompt "Enter LUKS passphrase" \
			--number-of-tries=3 --command="${ply_cmd_file}"
		res=$?
		run rm "${ply_cmd_file}"
		return ${res}
	else
		splashCmd 'verbose' >/dev/null &
		eval run ${1}
		res=$?
		if [ ${res} -eq 0 ]
		then
			splashCmd set_msg 'Disk unlocked.'
		fi
		return ${res}
	fi
}
```

FlyingWaffleDev replied:
I like this solution; it keeps everything strictly local.  Will push it after I test on my system.

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -2068,8 +2146,19 @@ openLUKS() {
+            crypt_filter_ret=$?
+            if [ "${PLYMOUTH}" = '1' ]
+            then
+                run rm -f "${ply_cmd_file}"
```
You can move it at the end of function near of others `run rm ...`

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -1226,16 +1228,29 @@ write_env_file() {
 	if [ "${CRYPT_SILENT}" = '1' ]
 	then
 		eval run ${1}
+	elif [ "${PLYMOUTH}" = '1' ] && plymouth_running && [ ${ask_pass} -eq 0 ]
```
BTW, why do you check every time if the plymouth is running? Is it really so unstable?

reagentoo replied:
I think you should move it near of `run plymouth` as in other cases:
`plymouth_running && run plymouth ...`

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -1553,6 +1571,71 @@ splash() {
 	return 0
 }
 
+splashCmd() {
```
Why not just implement an existing empty `splash()`?

FlyingWaffleDev replied:
The splash vs splashCmd stuff was mostly for trying to accommodate the deprecated fbsplash.
Rewritten in next commit to use existing splash() function, which also cleans up the changes a bit.

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -1553,6 +1571,71 @@ splash() {
+	esac
+}
+
+plymouth_running() {
```
```
plymouthRun() {
	run plymouth --ping 2>/dev/null || return $?
	run plymouth "${@}" 2>/dev/null
}
```

FlyingWaffleDev replied:
I really like this; in next commit.  Also resolves your previous comment.

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -1553,6 +1571,71 @@ splash() {
+	udevsettle
+
+	run plymouthd --mode=boot --attach-to-session \
+		--pid-file=/run/plymouth/pid || { PLYMOUTH_FAIL=1; return 1; }
```
```
run plymouthd --mode=boot --attach-to-session \
	--pid-file=/run/plymouth/pid
if [ $? -ne 0 ]
then
	bad_msg "Can't start plymouthd!"
	PLYMOUTH=0
	return 1
fi

plymouthRun --show-splash
if [ $? -eq 0 ]
then
	good_msg "Plymouth initialized"
fi
```

FlyingWaffleDev replied:
Added!

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -2055,7 +2139,11 @@ openLUKS() {
+						# plymouth password entry is passed through STDIN, requiring '--passphrase-fd 0 --batch'
+						# for newer gpg versions (>=2.1) '--pinentry-mode loopback' may also be required for the above
+						# '--no-tty' is included to prevent interruption of plymouth by any output
+						ply_cmd="gpg --logger-file /dev/null --quiet --passphrase-fd 0 --batch --no-tty --decrypt ${mntkey}${LUKS_KEY} | "
```
```
gpg_cmd="gpg --logger-file /dev/null --quiet --decrypt ${mntkey}${LUKS_KEY}"

if [ "${PLYMOUTH}" = '1' -a "${CRYPT_SILENT}" != '1' ]
then
	gpg_cmd="${gpg_cmd} --passphrase-fd 0 --batch --no-tty |"
else
	gpg_cmd="${gpg_cmd} |"
fi
```

FlyingWaffleDev replied:
Reread the gpg man page, looks like this should work fine.  Added!

FlyingWaffleDev replied:
I like how this reduces duplication, but it causes issues when the '--decrypt' option isn't last.  I've implemented a slight variation that works.

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -1226,16 +1228,29 @@ write_env_file() {
 }
 
 crypt_filter() {
+	[ -z "${LUKS_KEY}" -o "$(echo ${LUKS_KEY} | grep -o '.gpg$')" = ".gpg" ]
```
I forgot `LUKS_KEY` is local in `openLUKS()`. Although of course it will work as is. But you have to pass it by argument.
```
local luks_key="${1}"
[ -z "${luks_key}" -o "$(echo ${luks_key} | grep -o '.gpg$')" = ".gpg" ]
...
```

FlyingWaffleDev replied:
Will be in next commit/push.

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -2068,7 +2153,10 @@ openLUKS() {
 			# At this point, {header,key}file or not, we're ready!
-			crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+			gpg_cmd="${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+			good_msg "Using the following gpg command: ${gpg_cmd}" ${CRYPT_SILENT}
```
Why do you copy whole cryptsetup string to this variable? The content of `gpg_cmd` doesn't reflect his own name then. There may be no gpg command. If you want to show such message you can do this inside `crypt_filter()`:
`good_msg "Using the following decryption command: ${1}" ${CRYPT_SILENT}`

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -2068,7 +2153,10 @@ openLUKS() {
+			gpg_cmd="${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+			good_msg "Using the following gpg command: ${gpg_cmd}" ${CRYPT_SILENT}
+
+			crypt_filter "${LUKS_KEY}" "${gpg_cmd}"
```
May be it's better to pass AKS_PASS flag directly:
```
local ASK_PASS=0
local DEV_ERROR=0
local HEADER_ERROR=0 HEADERDEV_ERROR=0
local KEY_ERROR=0 KEYDEV_ERROR=0
...
	if [ -n "${LUKS_KEY}" ]
	then
		...
		if [ "$(echo ${LUKS_KEY} | grep -o '.gpg$')" = ".gpg" ]
		then
			...
		else
			ASK_PASS=1
			cryptsetup_options="${cryptsetup_options} -d ${mntkey}${LUKS_KEY}"
		fi
	...
	crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}" ${ASK_PASS}
```
(by the second arg)

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -2054,8 +2130,18 @@ openLUKS() {
 					else
 						[ -e /dev/tty ] && run mv /dev/tty /dev/tty.org
 						run mknod /dev/tty c 5 1
+						ASK_PASS=1
```
Why here? '1' means don't ask pwd.

FlyingWaffleDev replied:
I had assumed the default was to NOT ask, meaning 1 meant TO ask.

reagentoo replied:
By default is to ask.

reagentoo replied:
Ah, ok. My mistake. It's cannot be interpreted as return code. Then please add also additional ASK_PASS=1
```
if [ -n "${LUKS_KEY}" ]
then
...
else
ASK_PASS=1
```

reagentoo commented on defaults/initrd.scripts:
```diff
@@ -2068,7 +2154,7 @@ openLUKS() {
 
 			# At this point, {header,key}file or not, we're ready!
-			crypt_filter "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
+			crypt_filter "${ASK_PASS}" "${gpg_cmd}cryptsetup ${cryptsetup_options} luksOpen ${LUKS_DEVICE} ${LUKS_NAME}"
```
`ASK_PASS` should be at the second arg. Similar to `CRYPT_SILENT` flag for `{good,bad}_msg()`.

reagentoo commented:
I've just tested it. I think you shouldn't force user to enable `--all-ramdisk-modules` && `--firmware` options.
In my case, the ramdisk with both options after build takes 312M. (271M - without all modules, 14M - usual size)
My config: `sys-kernel/gentoo-kernel` + fimware packages:
```
$ equery b /lib/firmware/
 * Searching for /lib/firmware/ ... 
net-wireless/wireless-regdb-20220408 (/lib/firmware)
sys-firmware/intel-microcode-20220510_p20220508 (/lib/firmware)
sys-kernel/linux-firmware-20220509 (/lib/firmware)

$ du -h /lib/firmware/ | tail -n1
893M	/lib/firmware/
```
Plymouth can work without a kms. The user should be able to decide for himself if he needs a kms before the plymouth running. And if needed, he can decide how to copy the module and firmware - all at once or selectively.

FlyingWaffleDev committed:
Changed firmware and modules requirements to warnings.

Depending on system configuration and hardware, added firmware and
modules might not be required for Plymouth to function.

Signed-off-by: FlyingWaffle <flyingwaffle@pm.me>

FlyingWaffleDev commented:
Fair enough; I've changed the firmware/module requirements to a short warning message for each.

reagentoo commented:
> Fair enough; I've changed the firmware/module requirements to a short warning message for each.

Is it possible to copy only the necessary drm modules and firmware for the plymouth? Although perhaps this could be done in future commits.

FlyingWaffleDev commented:
> Is it possible to copy only the necessary drm modules and firmware for the plymouth? Although perhaps this could be done in future commits.

I'm still a little unsure how genkernel's default module loading/import works, but from looking through it, I think I can add generic loading for drm modules.  The firmware on the other hand is very system specific, and mostly user configured anyway.  So, I think I'll leave the warning for the firmware and add a commit for the module loading.  (Unless someone has a better idea.)

It's a little difficult for me to test though, because I have a unified kernel/initramfs efi archive and most of my modules are compiled into my kernel.

robbat2 commented:
> > Is it possible to copy only the necessary drm modules and firmware for the plymouth? Although perhaps this could be done in future commits.
> 
> I'm still a little unsure how genkernel's default module loading/import works, but from looking through it, I think I can add generic loading for drm modules. The firmware on the other hand is very system specific, and mostly user configured anyway. So, I think I'll leave the warning for the firmware and add a commit for the module loading. (Unless someone has a better idea.)

If the user is selecting the correct modules for their system, we could probe those modules to get their firmware lists as a first step.
E.g. `modinfo -F firmware amdgpu` gives me every possible firmware needed by the `amdgpu` module.  

Moving past that requires kernel patching, and my patches were rejected but they would have made it easier to identify exactly what firmware is loaded on a system:
https://lore.kernel.org/lkml/20191117234734.27101-1-robbat2@gentoo.org/T/#u

robbat2 commented:
What I've considered instead is that the firmware might belong in a separate /boot file, similar to the microcode, and get updated when the linux-firmware package is updated, but that's then messy for firmware from other packages, as well as easy of rollback if you get a bad firmware.

dracut's behavior I think is basically the `modinfo -F firmware` variant

reagentoo commented:
> dracut's behavior I think is basically the `modinfo -F firmware` variant

The guess is not far from the truth, but instead of the command, the C-approach is used:
https://github.com/dracutdevs/dracut/blob/f63f411d52df613936082d646ab072447b8b9d7f/src/install/dracut-install.c#L1382

I thought about another question - is it possible to optionally include modules to the image?
For example, to make an option `DRM="yes"`. And copy the modules from `/lib/modules/${kver}/kernel/drivers/gpu` if this option is enabled.

But actually I don't think it would be worth doing all of it in the context of this PR.

robbat2 commented:
> > dracut's behavior I think is basically the `modinfo -F firmware` variant

> The guess is not far from the truth, but instead of the command, the C-approach is used: https://github.com/dracutdevs/dracut/blob/f63f411d52df613936082d646ab072447b8b9d7f/src/install/dracut-install.c#L1382

yes, pretty much the same thing, just in C ;-)

> I thought about another question - is it possible to optionally include modules to the image? For example, to make an option `DRM="yes"`. And copy the modules from `/lib/modules/${kver}/kernel/drivers/gpu` if this option is enabled.
How does what you're asking for differ from the existing module system? 

Look for example at the `virtio` stuff in these files, as well as the outer loop of `HWOPTS` (all possible options) / `MY_HWOPTS` (user-selected options)
/usr/share/genkernel/defaults/modules_load
/usr/share/genkernel/defaults/initrd.scripts
/usr/share/genkernel/defaults/initrd.default

thesamesam commented:
Thank you for your patience! Reading over the comments, it looks like Robin's feedback was addressed, so I've gone ahead and merged it after a ping on IRC.

reagentoo commented:
@FlyingWaffleDev Hi.
Do you have any ideas on how to continue the Plymouth screen until the display manager starts? Plymouth only works while the genkernel's initrd image is loading. Since openrc starts, there is console output again without Plymouth.

FlyingWaffleDev commented:
> @FlyingWaffleDev Hi. Do you have any ideas on how to continue the Plymouth screen until the display manager starts? Plymouth only works while the genkernel's initrd image is loading. Since openrc starts, there is console output again without Plymouth.

I accomplish that with a combination of [silent boot](https://wiki.archlinux.org/title/Silent_boot), and the [plymouth-openrc-plugin](https://packages.gentoo.org/packages/sys-boot/plymouth-openrc-plugin). You also have to set openrc to [non-interactive.](https://wiki.gentoo.org/wiki/Plymouth#OpenRC)

It is a little finicky, though, and doesn't always show the splash. But a momentary blank screen is better than nothing.

reagentoo commented:
> I accomplish that with a combination of [silent boot](https://wiki.archlinux.org/title/Silent_boot), and the [plymouth-openrc-plugin](https://packages.gentoo.org/packages/sys-boot/plymouth-openrc-plugin). You also have to set openrc to [non-interactive.](https://wiki.gentoo.org/wiki/Plymouth#OpenRC)

I'm also using plymouth-openrc-plugin+rc_interactive="NO", but Plymouth never appears while openrc loading.
I'm not sure about "silent boot" mode. Because I want to see at least 1 line of the status at the Plymouth top.

FlyingWaffleDev commented:
> I'm also using plymouth-openrc-plugin+rc_interactive="NO", but Plymouth never appears while openrc loading.

Yeah, like I said, it's finicky. On my desktop I had it installed but not working at all, then I had to reinstall after a drive failure and on the fresh install it worked fine (mostly).
Also, depending on your plymouth theme, the plugin should let it show openrc and other boot status updates even with silent boot on.
